src/Tests/Microsoft.Diagnostics.Monitoring.Tool.FunctionalTests/CollectionRuleAndApiTests.cs

```diff
@@ -0,0 +1,236 @@
+                ApiClient apiClient = new(_outputHelper, httpClient);
+
+                await CaptureTrace(appRunner, apiClient);
+
+                Task ruleStartedTask = toolRunner.WaitForCollectionRuleActionsCompletedAsync(DefaultRuleName);
+
+                await ValidateAspNetTriggerCollected(
+                    ruleStartedTask,
+                    apiClient,
+                    ClientBaseUrl,
+                    urlPaths,
+                    ExpectedFilePath,
+                    ExpectedFileContent);
+
+                await appRunner.SendCommandAsync(TestAppScenarios.AspNetSlow.Commands.Continue);
+            });
+        }
+
+        private async Task ValidateAspNetTriggerCollected(Task ruleTask, ApiClient client, string hostName, string[] paths, string expectedFilePath, string expectedFileContent)
+        {
+            await ApiCallHelper(hostName, paths, client);
+
+            await ruleTask;
+            Assert.True(ruleTask.IsCompleted);
+
+            Assert.True(File.Exists(expectedFilePath));
+            Assert.Equal(expectedFileContent, File.ReadAllText(expectedFilePath));
+
+            File.Delete(expectedFilePath);
+        }
+
+        private AppRunner SetUpAppRunner(DiagnosticPortConnectionMode appConnectionMode, string diagnosticPortPath)
+        {
+            AppRunner appRunner = new(_outputHelper, Assembly.GetExecutingAssembly());
+            appRunner.ConnectionMode = appConnectionMode;
+            appRunner.DiagnosticPortPath = diagnosticPortPath;
+            appRunner.ScenarioName = TestAppScenarios.AspNetSlow.Name;
+            appRunner.Environment.Add(AspNetUrlsKey, AppUrl);
+
+            return appRunner;
+        }
+
+        private MonitorCollectRunner SetUpToolRunner(DiagnosticPortConnectionMode mode, string diagnosticPortPath)
+        {
+            MonitorCollectRunner toolRunner = new(_outputHelper);
+            toolRunner.ConnectionModeViaCommandLine = mode;
+            toolRunner.DiagnosticPortPath = diagnosticPortPath;
+            toolRunner.DisableAuthentication = true;
+
+            return toolRunner;
+        }
+
+        private async Task ApiCallHelper(string hostName, string[] paths, ApiClient client)
+        {
+            foreach (string path in paths)
+            {
+                string url = hostName + path;
+                _ = await client.ApiCall(url);
+
+                await Task.Delay(TimeSpan.FromMilliseconds(200));
+            }
+        }
+
+        private async Task<ResponseStreamHolder> CaptureTrace(AppRunner runner, ApiClient client)
+        {
+            int processId = await runner.ProcessIdTask;
+
+            int retryCounter = 0;
+
+            // Repeatedly check if dotnet-monitor is detecting our process; due to timing issues,
+            // tests were sometimes failing due to the target process not being found when collecting a trace.
+            while (retryCounter < 5)
```

Comment:
Why is this needed? The `GetProcessesWithRetryAsync` call already takes care of retries.